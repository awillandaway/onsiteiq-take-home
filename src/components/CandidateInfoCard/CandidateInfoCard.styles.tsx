import { ButtonWrapper } from 'components/shared/Button/Button.styles';
import { onsiteIqPurple } from 'constants/colors';
import styled, { css } from 'styled-components';

export const CandidateInfoCardWrapper = styled.div``;

export const PhotoAndName = styled.div`
  display: flex;
`;

export const Name = styled.div`
  font-size: 32px;
`;

export const Status = styled.div`
  font-size: 20px;
`;

export const NameAndStatus = styled.div`
  display: flex;
  flex-direction: column;
  margin-left: 16px;
`;

export const SectionHeader = styled.div`
  text-transform: uppercase;
  font-size: 10px;
  margin: 24px 0 12px;
  padding: 0 4px;
`;

export const NotesTextArea = styled.textarea`
  ${(props) =>
    props.disabled &&
    css`
      color: white;
    `}
  margin-top: 8px;
  resize: none;
  height: 6rem;
  outline: none;
  border-radius: 2px;
  &:focus {
    box-shadow: 0px 0px 0 2px ${onsiteIqPurple};
  }
`;

export const ButtonSection = styled.div`
  display: flex;
  margin-top: 8px;

  ${ButtonWrapper} {
    flex: 1;
    &:not(:last-of-type) {
      margin-right: 8px;
    }
  }
`;
