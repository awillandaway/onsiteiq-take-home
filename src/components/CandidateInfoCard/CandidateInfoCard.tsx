import { Button } from 'components/shared/Button/Button';
import { Card } from 'components/shared/Card/Card';
import { useState } from 'react';
import type { ExtendedCandidate } from 'services/CandidateService';
import type { Candidate } from 'types/Candidate';
import { capitalizeFirstLetter } from 'utils/StringUtils';
import {
  ButtonSection,
  Name,
  NameAndStatus,
  NotesTextArea,
  PhotoAndName,
  SectionHeader,
  Status,
} from './CandidateInfoCard.styles';
import { CardField } from './CardField/CardField';

interface CandidateInfoCardProps {
  candidate: Candidate | ExtendedCandidate;
  onApproveCandidate: (candidate: Candidate, notes?: string) => void;
  onRejectCandidate: (candidate: Candidate, notes?: string) => void;
}

export const CandidateInfoCard = ({ candidate, onApproveCandidate, onRejectCandidate }: CandidateInfoCardProps) => {
  const [notes, setNotes] = useState(candidate.notes);

  const formatCandidateName = () => {
    const { title, first, last } = candidate.name;
    return `${title} ${first} ${last}`; // should we show title?
  };

  const formatStreet = () => {
    const { number, name } = candidate.location.street;
    return `${number} ${name}`;
  };

  const formatCityState = () => {
    const { city, state, postcode } = candidate.location;
    return `${city}, ${state} ${postcode}`;
  };

  const registeredDate = new Date(candidate.registered.date);

  const getBackgroundColor = () => {
    switch (candidate.status) {
      case 'approved':
        return '#0c1a0c';
      case 'rejected':
        return '#2a0606';
      case 'not_yet_reviewed':
      default:
        return undefined;
    }
  };

  return (
    <Card backgroundColor={getBackgroundColor()}>
      <PhotoAndName>
        <img src={candidate.picture.large} alt="candidate" />

        <NameAndStatus>
          <Name>{formatCandidateName()}</Name>
          {candidate.status !== 'not_yet_reviewed' && (
            <Status>Status: {capitalizeFirstLetter(candidate.status)}</Status>
          )}
        </NameAndStatus>
      </PhotoAndName>

      <SectionHeader>Basic Info</SectionHeader>
      <CardField label="Registered">{registeredDate.toLocaleDateString()}</CardField>
      <CardField label="Gender">{capitalizeFirstLetter(candidate.gender)}</CardField>
      <CardField label="Age">{candidate.dob.age}</CardField>

      <SectionHeader>Contact Info</SectionHeader>
      <CardField label="Email">{candidate.email}</CardField>
      <CardField label="Address">
        <div>
          <div>{formatStreet()}</div>
          <div>{formatCityState()}</div>
          <div>{candidate.location.country}</div>
        </div>
      </CardField>

      <CardField label="Phone">Phone: {candidate.phone}</CardField>
      <CardField label="Cell">Cell: {candidate.cell}</CardField>

      <SectionHeader>Notes</SectionHeader>
      <NotesTextArea
        autoFocus
        placeholder="Notes about the candidate"
        defaultValue={candidate.notes}
        disabled={candidate.status !== 'not_yet_reviewed'}
        onChange={(event) => setNotes(event.target.value)}
      />
      <ButtonSection>
        {candidate.status === 'not_yet_reviewed' ? (
          <>
            <Button text="Approve" variant="primary" onClick={() => onApproveCandidate(candidate, notes)} />
            <Button text="Reject" variant="danger" onClick={() => onRejectCandidate(candidate, notes)} />
          </>
        ) : (
          <Button text="Edit Review" />
        )}
      </ButtonSection>
    </Card>
  );
};
