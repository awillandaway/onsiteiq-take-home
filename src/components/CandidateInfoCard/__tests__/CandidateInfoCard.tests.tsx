import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import type { Candidate } from 'types/Candidate';
import { CandidateInfoCard, getBackgroundColor } from '../CandidateInfoCard';

const mockCandidate: Candidate = {
  gender: 'male',
  name: {
    title: 'Mr',
    first: 'John',
    last: 'Smith',
  },
  location: {
    street: {
      number: 1,
      name: 'Test Street',
    },
    city: 'New York',
    state: 'New York',
    country: 'United States',
    postcode: 11111,
    coordinates: {
      latitude: 'test-latitude',
      longitude: 'test-longitude',
    },
    timezone: {
      offset: '-5:00',
      description: 'US East',
    },
  },
  email: 'john.smith@test.com',
  dob: {
    date: '1955-11-04T21:17:18.030Z',
    age: 67,
  },
  registered: {
    date: '2017-10-15T15:36:03.501Z',
    age: 5,
  },
  phone: '(123) 456-7890',
  cell: '(111) 111-1111',
  id: {
    name: 'drivers license',
    value: '11111111111',
  },
  picture: {
    large: 'https://randomuser.me/api/portraits/men/76.jpg',
    medium: 'https://randomuser.me/api/portraits/med/men/76.jpg',
    thumbnail: 'https://randomuser.me/api/portraits/thumb/men/76.jpg',
  },
  nat: 'US',
  status: 'not_yet_reviewed',
};

const mockApprovedCandidate: Candidate = { ...mockCandidate, status: 'approved', notes: 'this candidate is great' };
const mockRejectedCandidate: Candidate = { ...mockCandidate, status: 'rejected' };

const mockOnApproveCandidate = jest.fn();
const mockOnRejectCandidate = jest.fn();
const mockOnClickEditReview = jest.fn();

const renderCandidateInfoCard = (candidate: Candidate = mockCandidate) =>
  render(
    <CandidateInfoCard
      candidate={candidate}
      onApproveCandidate={mockOnApproveCandidate}
      onRejectCandidate={mockOnRejectCandidate}
      onClickEditReview={mockOnClickEditReview}
    />,
  );

describe('CandidateInfoCard', () => {
  it('Renders correctly', () => {
    const card = renderer
      .create(
        <CandidateInfoCard
          candidate={mockCandidate}
          onApproveCandidate={mockOnApproveCandidate}
          onRejectCandidate={mockOnRejectCandidate}
        />,
      )
      .toJSON();
    expect(card).toMatchSnapshot();
  });

  it('Correctly gets background color', () => {
    expect(getBackgroundColor(mockCandidate)).toBe(undefined);
    expect(getBackgroundColor(mockApprovedCandidate)).toBe('#0c1a0c');
    expect(getBackgroundColor(mockRejectedCandidate)).toBe('#2a0606');
  });

  it('Does not renders candidate status when candidate is not yet reviewed', () => {
    const { queryByText } = renderCandidateInfoCard();

    expect(queryByText((text) => text.includes('Status'))).toBeNull();
  });

  it('Renders candidate status when candidate is approved', () => {
    const { getByText, queryByText } = renderCandidateInfoCard(mockApprovedCandidate);

    expect(getByText('Status: Approved')).toBeVisible();
    expect(queryByText('Status: Rejected')).toBeNull();
  });

  it('Renders candidate status when candidate is rejected', () => {
    const { getByText, queryByText } = renderCandidateInfoCard(mockRejectedCandidate);

    expect(getByText('Status: Rejected')).toBeVisible();
    expect(queryByText('Status: Approved')).toBeNull();
  });

  it('Renders approve and reject buttons when candidate is not yet reviewed', () => {
    const { getByRole } = renderCandidateInfoCard();

    expect(getByRole('button', { name: 'Approve' })).toBeVisible();
    expect(getByRole('button', { name: 'Reject' })).toBeVisible();
  });

  it('Renders edit review button when candidate is approved and disables notes textbox', () => {
    const { getByRole, queryByRole } = renderCandidateInfoCard(mockApprovedCandidate);

    expect(getByRole('textbox')).toBeDisabled();

    expect(getByRole('button', { name: 'Edit Review' })).toBeVisible();
    expect(queryByRole('button', { name: 'Approve' })).toBeNull();
    expect(queryByRole('button', { name: 'Reject' })).toBeNull();
  });

  it('Renders edit review button when candidate is rejected and disables notes textbox', () => {
    const { getByRole, queryByRole } = renderCandidateInfoCard(mockRejectedCandidate);

    expect(getByRole('textbox')).toBeDisabled();

    expect(getByRole('button', { name: 'Edit Review' })).toBeVisible();
    expect(queryByRole('button', { name: 'Approve' })).toBeNull();
    expect(queryByRole('button', { name: 'Reject' })).toBeNull();
  });

  it('Clicking Approve calls onApproveCandidate with the correct parameters', () => {
    const { getByRole } = renderCandidateInfoCard();
    const notesTextArea = getByRole('textbox') as HTMLTextAreaElement;
    fireEvent.change(notesTextArea, { target: { value: 'test notes' } });

    expect(notesTextArea.value).toBe('test notes');
    fireEvent.click(getByRole('button', { name: 'Approve' }));

    expect(mockOnApproveCandidate).toHaveBeenCalledTimes(1);
    expect(mockOnApproveCandidate).toHaveBeenCalledWith(mockCandidate, 'test notes');
  });

  it('Clicking Reject calls onRejectCandidate with the correct parameters', () => {
    const { getByRole } = renderCandidateInfoCard();
    const notesTextArea = getByRole('textbox') as HTMLTextAreaElement;
    fireEvent.change(notesTextArea, { target: { value: 'test notes 2' } });

    expect(notesTextArea.value).toBe('test notes 2');
    fireEvent.click(getByRole('button', { name: 'Reject' }));

    expect(mockOnRejectCandidate).toHaveBeenCalledTimes(1);
    expect(mockOnRejectCandidate).toHaveBeenCalledWith(mockCandidate, 'test notes 2');
  });

  it('Clicking Edit Review calls onClickEditReview with the correct parameters', () => {
    const { getByRole } = renderCandidateInfoCard(mockApprovedCandidate);

    fireEvent.click(getByRole('button', { name: 'Edit Review' }));

    expect(mockOnClickEditReview).toHaveBeenCalledTimes(1);
    expect(mockOnClickEditReview).toHaveBeenCalledWith(mockApprovedCandidate, 'this candidate is great');
  });
});
