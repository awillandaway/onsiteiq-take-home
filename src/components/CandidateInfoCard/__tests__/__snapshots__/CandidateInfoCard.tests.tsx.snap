// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CandidateInfoCard Renders correctly 1`] = `
<div
  className="sc-gswNZR fwyvqt"
>
  <div
    className="sc-hLBbgP gJsMjZ"
  >
    <img
      alt="candidate"
      src="https://randomuser.me/api/portraits/men/76.jpg"
    />
    <div
      className="sc-gKPRtg jGdynd"
    >
      <div
        className="sc-eDvSVe QHMfZ"
      >
        Mr John Smith
      </div>
    </div>
  </div>
  <div
    className="sc-iBYQkv hmotmI"
  >
    Basic Info
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Registered
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      10/15/2017
    </div>
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Gender
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      Male
    </div>
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Age
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      67
    </div>
  </div>
  <div
    className="sc-iBYQkv hmotmI"
  >
    Contact Info
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Email
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      john.smith@test.com
    </div>
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Address
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      <div>
        <div>
          1 Test Street
        </div>
        <div>
          New York, New York 11111
        </div>
        <div>
          United States
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Phone
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      Phone: 
      (123) 456-7890
    </div>
  </div>
  <div
    className="sc-jrcTuL dzWBFN"
  >
    <div
      className="sc-kDvujY eutWaG"
    >
      Cell
    </div>
    <div
      className="sc-ipEyDJ eSJufq"
    >
      Cell: 
      (111) 111-1111
    </div>
  </div>
  <div
    className="sc-iBYQkv hmotmI"
  >
    Notes
  </div>
  <textarea
    aria-label="notes-input"
    autoFocus={true}
    className="sc-ftTHYK CzUZx"
    disabled={false}
    onChange={[Function]}
    placeholder="Notes about the candidate"
  />
  <div
    className="sc-pyfCe frezrP"
  >
    <button
      className="sc-bcXHqe eftqLE"
      onClick={[Function]}
      type="button"
    >
      Approve
    </button>
    <button
      className="sc-bcXHqe edihgN"
      onClick={[Function]}
      type="button"
    >
      Reject
    </button>
  </div>
</div>
`;
